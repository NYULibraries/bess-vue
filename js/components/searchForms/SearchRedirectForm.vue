<template lang="html">
  <form @submit.prevent="openSearch" :class="`${searchKey}-search-form`">
    <div class="bobcat_embed_search_field">
      <span class="bobcat_embed_"><label :for="`${searchKey}-query`">Search for</label>
        <input
          :aria-label="inputAriaLabel"
          type="text"
          :id="`${searchKey}-query`"
          class="bobcat_embed_searchbox_textfield"
          v-model="search"
          :key="searchKey"
        >
      </span>

      <span class="bobat_embed_searchbox_submit_container">
        <input aria-label="Search" class="bobcat_embed_searchbox_submit" name="Submit" type="submit" value="GO">
      </span>
    </div>
  </form>
</template>

<script>

export default {
  name: "guides-search-form",
  data() {
    return {
      search: '',
    };
  },
  props: [
    'searchKey',
    'searchFunction',
    'searchEngineProps',
    'inputAriaLabel',
  ],
  methods: {
    openSearch() {
      window.open(this.searchFunction(this.searchValues));
    }
  },
  computed: {
    searchValues() {
      return {
        search: this.search,
        ...this.searchEngineProps,
      };
    },
  },
};
</script>

<style lang="css">
</style>
