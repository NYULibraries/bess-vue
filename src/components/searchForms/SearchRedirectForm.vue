<template lang="html">
  <form
    :class="`tab-${ searchKey }-search-form`"
    @submit.prevent="openSearch"
  >
    <div class="bobcat_embed_search_field">
      <span class="bobcat_embed_"><label :for="`tab-${ searchKey }-query`">Search for </label>
        <input
          :id="`tab-${ searchKey }-query`"
          :key="searchKey"
          v-model="search"
          :aria-label="inputAriaLabel"
          type="text"
          class="bobcat_embed_searchbox_textfield"
          :placeholder="searchEngineProps.placeholder"
          :aria-describedby="searchEngineProps.placeholder"
        >
      </span>

      <!-- Dropdown for selecting search scope -->
      <select
        :id="`tab-${ searchKey }-scope`"
        v-model="selectedScope"
        class="bobcat_embed_select_value"
        aria-label="Select search scope"
      >
        <option
          v-for="( config, scopeValue ) in scopesConfig"
          :key="scopeValue"
          :value="scopeValue"
        >
          {{ config.label }}
        </option>
      </select>

      <span class="bobat_embed_searchbox_submit_container">
        <input
          aria-label="Search"
          class="bobcat_embed_searchbox_submit"
          name="Submit"
          type="submit"
          value="GO"
        >
      </span>
    </div>
  </form>
</template>

<script>
export default {
    name : 'GuidesSearchForm',
    props: [
        'searchKey',
        'searchFunction',
        'searchEngineProps',
        'inputAriaLabel',
    ],
    data() {
        return {
            search       : '',
            selectedScope: 'CI_NYU_CONSORTIA',
        };
    },
    computed: {
        searchValues() {
            return {
                search: this.search,
                ...this.searchEngineProps,
                scope : this.selectedScope,
            };
        },
    },
    watch: {
        selectedScope( scope ) {
            this.updatePlaceholder( scope );
        },
    },
    created() {
        this.updatePlaceholder( this.selectedScope );
    },
    methods: {
        openSearch() {
            console.log( 'searchvalues:', this.searchValues  );
            window.open( this.searchFunction( this.searchValues ) );
        },
        updatePlaceholder( scope ) {
            const newPlaceholder = this.scopesConfig[scope]?.placeholder || '';
            if ( this.searchEngineProps.placeholder !== newPlaceholder ) {
                this.$emit( 'update:searchEngineProps', {
                    ...this.searchEngineProps,
                    placeholder: newPlaceholder,
                } );
            }
        },
    },
};
</script>

<style lang="css">
</style>
