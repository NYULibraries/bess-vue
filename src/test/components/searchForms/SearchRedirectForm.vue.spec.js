import { afterEach, beforeEach, describe, expect, it } from 'vitest';
import { vi } from 'vitest';
import SearchRedirectForm from '../../../components/searchForms/SearchRedirectForm.vue';
import { shallowMount } from '@vue/test-utils';

const inputAriaLabel = 'Search in library guides';
const searchFunctionSpy = vi.fn( () => 0 );

const props = {
    searchKey        : 'guides',
    searchFunction   : searchFunctionSpy,
    searchEngineProps: {
        guidesUrl: 'http://guides.library.edu',
        prop1    : 'prop1',
        prop2    : 'prop2',
    },
    inputAriaLabel,
    ui: {
        searchScopeDropdown: {
            defaultOption: 'Library catalog',
            options      : {
                'Library catalog': {
                    label      : 'Library catalog',
                    placeholder: 'Search for books, articles, etc.',
                },
            },
        },
    },
};

describe( 'SearchRedirectForm', () => {
    let wrapper;
    beforeEach( () => {
        wrapper = shallowMount( SearchRedirectForm, {
            props,
            attachToDocument: false,
        } );
    } );

    afterEach( () => {
        searchFunctionSpy.mockReset();
    } );

    it( 'is a Vue instance', () => {
        expect( wrapper ).toBeTruthy();
    } );

    it( 'has the correct HTML', () => {
        expect( wrapper.html() ).toMatchSnapshot();
    } );

    describe( 'props', () => {
        it( 'includes searchKey, searchFunction, inputAriaLabel, and ui as props', () => {
            expect( Object.keys( wrapper.props() ).length ).toEqual( 5 );
            expect( wrapper.props().searchKey ).toEqual( props.searchKey );
            expect( wrapper.props().searchFunction ).toEqual( props.searchFunction );
            expect( wrapper.props().searchEngineProps ).toEqual( props.searchEngineProps );
            expect( wrapper.props().inputAriaLabel ).toEqual( props.inputAriaLabel );
            expect( wrapper.props().ui ).toEqual( props.ui );
        } );
    } );

    describe( 'data', () => {
        it( 'initializes with empty string for \'search\'', () => {
            expect( wrapper.vm.search ).toBe( '' );
        } );

        it( 'initializes selectedSearchScope with correct default scope', () => {
            expect( wrapper.vm.selectedSearchScope ).toBe( props.ui.searchScopeDropdown.defaultOption );
        } );
    } );

    describe( 'methods', () => {
        const search = 'this is a search term';

        // Suppress error "Error: Not implemented: window.open", which doesn't
        // cause the test to fail but does make test output harder to read.
        vi.stubGlobal( 'open', vi.fn() );

        describe( 'openSearch', () => {
            beforeEach( () => {
                wrapper.setData( {
                    search,
                    selectedSearchScope: 'Library catalog',
                } );

                vi.spyOn( window, 'open' );
                wrapper.vm.openSearch();
            } );

            it( 'should call open window with searchFunction', () => {
                expect( window.open ).toHaveBeenCalled();
                expect( searchFunctionSpy ).toHaveBeenCalled();
                expect( searchFunctionSpy ).toHaveBeenCalledWith( {
                    search,
                    ...props.searchEngineProps,
                    scope: 'Library catalog',
                } );
            } );
        } );
    } );

    describe( 'computed', () => {
        describe( 'searchValues', () => {
            const search = 'this is a search term';
            beforeEach( () => {
                wrapper.setData( {
                    search,
                    selectedSearchScope: 'Library catalog',
                } );
            } );
            it( 'builds required parameters for guidesSearch as POJO from this.search (data) and this.engineValues (computed)', () => {
                expect( wrapper.vm.searchValues ).toEqual( {
                    search,
                    ...props.searchEngineProps,
                    scope: 'Library catalog',
                } );
            } );
        } );
    } );

    describe( 'shallow render', () => {
        it( 'has a form', () => {
            expect( wrapper.find( 'form' ).exists() ).toBeTruthy();
        } );

        it( 'form a div with class "bobcat_embed_search_field"', () => {
            expect( wrapper.find( 'form' ).find( 'div.bobcat_embed_search_field' ).exists() ).toBeTruthy();
        } );

        describe( 'the input', () => {
            let input;
            beforeEach( () => {
                input = wrapper.find( 'input[type=text]' );
            } );

            it( 'exists', () => {
                expect( input.exists() ).toBeTruthy();
            } );

            it( 'has an aria-label based on inputAriaLabel', () => {
                expect( input.attributes( 'aria-label' ) ).toEqual( props.inputAriaLabel );
            } );

            it( 'has an id and corresponding label', () => {
                expect( input.attributes( 'id' ) ).toEqual( `tab-${ props.searchKey }-query` );
                expect( wrapper.findAll( `label[for="tab-${ props.searchKey }-query"]` ).length ).toEqual( 1 );
            } );

            it( 'has class "bobcat_embed_searchbox_textfield"', () => {
                expect( input.classes( 'bobcat_embed_searchbox_textfield' ) ).toBeTruthy();
            } );

            it( 'is wrapped in a span with class "bobcat_embed_"', () => {
                const parentEl = input.element.parentElement;
                const klasses = Array.from( parentEl.classList );
                expect( klasses.indexOf( 'bobcat_embed_' ) > -1 ).toBeTruthy();
            } );

            it( 'should be bound to \'search\' in data', () => {
                input.setValue( 'typing' );
                expect( wrapper.vm.search ).toEqual( 'typing' );
            } );

            it( 'has placeholder text, if defined', () => {
                const expectedPlaceholder = props.ui.searchScopeDropdown.options[
                    props.ui.searchScopeDropdown.defaultOption
                ].placeholder;
                expect( input.attributes( 'placeholder' ) ).toBe( expectedPlaceholder );
            } );

            it( 'has aria-describedby according to placeholder text, if defined', () => {
                const expectedPlaceholder = props.ui.searchScopeDropdown.options[
                    props.ui.searchScopeDropdown.defaultOption
                ].placeholder;
                expect( input.attributes( 'aria-describedby' ) ).toBe( expectedPlaceholder );
            } );
        } );

        it( 'includes a submit button', () => {
            expect( wrapper.findAll( 'input[type=submit]' ).length ).toEqual( 1 );
        } );

        describe( 'form submit', () => {
            // NOTE: `wrapper` is not yet defined, so we can't create the spy.
            //       We'll have to do it in the `beforeEach` callback.
            let openSearchSpy;

            afterEach( () => {
                openSearchSpy.mockReset();
            } );

            beforeEach( () => {
                // NOTE: `wrapper` isn't defined yet in the outer `describe` scope,
                //       so we have to repeatedly re-create it here.
                openSearchSpy = vi.spyOn( wrapper.vm, 'openSearch' );
                wrapper.find( 'form' ).trigger( 'submit' );
            } );

            it( 'should trigger openSearch method', () => {
                expect( openSearchSpy ).toHaveBeenCalled();
            } );
        } );
    } );
} );
