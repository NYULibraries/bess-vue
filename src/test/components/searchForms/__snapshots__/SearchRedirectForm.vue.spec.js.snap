// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchRedirectForm > has the correct HTML 1`] = `
"<form class="tab-guides-search-form">
  <div class="bobcat_embed_search_field"><span class="bobcat_embed_"><label for="tab-guides-query">Search for </label><input id="tab-guides-query" aria-label="Search in library guides" type="text" class="bobcat_embed_searchbox_textfield" placeholder="Placeholder text" aria-describedby="Placeholder text"></span><span class="bobat_embed_searchbox_submit_container"><input aria-label="Search" class="bobcat_embed_searchbox_submit" name="Submit" type="submit" value="GO"></span></div>
</form>"
`;
